zero_bin/prover/src/lib.rs

```diff
@@ -115,91 +128,85 @@ impl BlockProverInput {
+    runtime: &Runtime,
+    previous_proof: Option<GeneratedBlockProof>,
+    save_inputs_on_error: bool,
+    proof_output_dir: Option<PathBuf>,
+) -> Result<Vec<(BlockNumber, Option<GeneratedBlockProof>)>> {
+    let mut prev: Option<BoxFuture<Result<GeneratedBlockProof>>> =
+        previous_proof.map(|proof| Box::pin(futures::future::ok(proof)) as BoxFuture<_>);
+
+    let mut results = FuturesOrdered::new();
+    for block_prover_input in block_prover_inputs {
+        let (tx, rx) = oneshot::channel::<GeneratedBlockProof>();
+        let proof_output_dir = proof_output_dir.clone();
+        let previos_block_proof = prev.take();
+        let fut = async move {
+            // Get the prover input data from the external source (e.g. Erigon node).
+            let block = block_prover_input.await?;
+            let block_number = block.get_block_number();
+            info!("Proving block {block_number}");
+
+            // Prove the block
+            let block_proof = block
+                .prove(runtime, previos_block_proof, save_inputs_on_error)
+                .then(move |proof| async move {
+                    let proof = proof?;
+                    let block_number = proof.b_height;
+
+                    // Write latest generated proof to disk if proof_output_dir is provided
+                    let return_proof: Option<GeneratedBlockProof> = if proof_output_dir.is_some() {
+                        write_proof(proof_output_dir, &proof).await?;
+                        None
+                    } else {
+                        Some(proof.clone())
+                    };
+
+                    if tx.send(proof).is_err() {
+                        anyhow::bail!("Failed to send proof");
+                    }
+
+                    Ok((block_number, return_proof))
+                })
+                .await?;
 
-        results.try_collect().await
+            Ok(block_proof)
+        }
+        .boxed();
+        prev = Some(Box::pin(rx.map_err(anyhow::Error::new)));
+        results.push_back(fut);
     }
 
-    /// Write the proof to the disk (if `output_dir` is provided) or stdout.
-    pub(crate) async fn write_proof(
-        output_dir: Option<PathBuf>,
-        proof: &GeneratedBlockProof,
-    ) -> Result<()> {
-        let proof_serialized = serde_json::to_vec(proof)?;
-        let block_proof_file_path =
-            output_dir.map(|path| generate_block_proof_file_name(&path.to_str(), proof.b_height));
-        match block_proof_file_path {
-            Some(p) => {
-                if let Some(parent) = p.parent() {
-                    tokio::fs::create_dir_all(parent).await?;
-                }
-
-                let mut f = tokio::fs::File::create(p).await?;
-                f.write_all(&proof_serialized)
-                    .await
-                    .context("Failed to write proof to disk")
+    results.try_collect().await
+}
+
+/// Write the proof to the disk (if `output_dir` is provided) or stdout.
```

Comment:
leftover comment technically unrelated to the changes, but `write_proof` will never print to `stdout` given that it's called only if `output_dir` is `Some`, so maybe worth cleaning up that function + comment?